### PR TITLE
Module 1

### PR DESCRIPTION
**Pull request by Alejandro-Andreo**
# PROBLEMAS
Por el tema del _rate limit_ de la API, me he limitado a realizar los gráficos únicamente con los datos de 1950 a 1960, es decir, 10 años.
Perdón por el retraso, tenia un pequeño error en el código y hasta que me he dado cuenta de que hacia bastantes más _requests_ de las que debería, han pasado demasiados días.
# GRÁFICOS
He hecho tres gráficos, uno para cada variable. En cada gráfico se añade los datos de cada ciudad durante de 10 años con su media y dispersión. Como todas las ciudades tienen más o menos los mismos resultados, se ven muy solapados.

temperature_2m

soil_moisture

Precipitation_sum

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -10,12 +10,14 @@ python = "^3.11"
 pandas = "^2.0.3"
 matplotlib = "^3.7.2"
 numpy = "^1.25.2"
-sklearn = "^0.0.post7"
+# sklearn = "^0.0.post7"
```
Comment (pabaldonedo) [inline pyproject.toml]: Do not leave unnecessary lines commented, just delete them

File: src/module_1/annual_data.csv
Comment (pabaldonedo) [inline src/module_1/annual_data.csv:1]: Data is seldom recommended to add it to the repo. In this case, why would you want to add it? We have the code to download it and regenerate it at will

File: src/module_1/module_1_meteo_api.py [hunk 2]
```diff
@@ -1,7 +1,248 @@
-""" This is a dummy example """
+import pandas as pd
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: Good that you grouped the "Imports" and then the "from import". TIP: sort the imports alphabetically. More: https://stackoverflow.com/questions/20762662/whats-the-correct-way-to-sort-python-import-x-and-from-x-import-y-statement

File: src/module_1/module_1_meteo_api.py [hunk 3]
```diff
@@ -1,7 +1,248 @@
...
+    "FGOALS_f3_H",
+    "HiRAM_SIT_HR",
+    "MRI_AGCM3_2_S",
+    "EC_Earth3P_HR",
+    "MPI_ESM1_2_XR",
+    "NICAM16_8S",
+]
+MAX_CALL_ATTEMPTS = 50
+
+# Logging configuration
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: self evident comment

File: src/module_1/module_1_meteo_api.py [hunk 4]
```diff
@@ -1,7 +1,248 @@
...
+    list_data_cities = []
+    for city in COORDINATES.keys():
+        logging.info(f"Getting data for {city}")
+        list_data_variables = []
+        for variable in VARIABLES:
+            logging.info(f"Getting data for {variable} in {city}")
+            for model in MODELS:
+                logging.info(f"Getting data for {model} in {city}")
+                data = get_data_meteo_api(
+                    city, "1950-01-01", "1960-12-31", variable, model
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: Probably cleaner and more flexible if you define the dates in variables

File: src/module_1/module_1_meteo_api.py [hunk 5]
```diff
@@ -1,7 +1,248 @@
...
+    plt.show()
+
 
 def main():
-    raise NotImplementedError
+    list_data_cities = []
+    for city in COORDINATES.keys():
+        logging.info(f"Getting data for {city}")
+        list_data_variables = []
+        for variable in VARIABLES:
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:244]: The API accepts several variables and models at once. This would reduce the amount of calls and boost efficiency

File: src/module_1/module_1_meteo_api.py [hunk 6]
```diff
@@ -1,7 +1,248 @@
...
+                    f"Rate limit exceeded. Waiting for {cool_off_time} seconds."
+                )
+                logging.info(response.text)
+                time.sleep(cool_off_time)
+                backoff_time *= 2
+            else:
+                raise Exception(
+                    f"API request failed with status code {response.status_code}"
+                )
+        except Exception as e:
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: Not good practice to cover ALL exceptions. Be more granular. In this [response](https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module) of Stackoverflow you can find a good summary of the exceptions to cover.

File: src/module_1/module_1_meteo_api.py [hunk 7]
```diff
@@ -1,7 +1,248 @@
...
+    Makes an API call with given parameters, handling rate limits and other errors.
+    :param params: Dictionary with the parameters to be passed to the API call.
+    :return: The API response if successful, None otherwise.
+    """
+    call_attempts = 0
+    backoff_time = 60
+    while call_attempts < MAX_CALL_ATTEMPTS:
+        try:
+            response = requests.get(API_URL, params=params)
+            if response.status_code == 200:
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: to make things more redable you can use `response.raise_for_status` and tackle non 200 statuses in except blocks

File: src/module_1/module_1_meteo_api.py [hunk 8]
```diff
@@ -1,7 +1,248 @@
...
+                raise Exception(
+                    f"API request failed with status code {response.status_code}"
+                )
+        except Exception as e:
+            logging.info(f"API request failed with exception {e}")
+            break
+        call_attempts += 1
+        logging.info(f"Call attempts: {call_attempts}")
+    logging.info("Max number of call attempts reached. Returning empty result")
+    return response
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: you could arrive here with a non-empty result. If you reach the max attemps and the response does not pass the validate_response the code will reach it here with non empty response

File: src/module_1/module_1_meteo_api.py [hunk 9]
```diff
@@ -1,7 +1,248 @@
...
+        return False
+
+    return True
+
+
+def call_api(params: dict):
+    """
+    Makes an API call with given parameters, handling rate limits and other errors.
+    :param params: Dictionary with the parameters to be passed to the API call.
+    :return: The API response if successful, None otherwise.
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:66]: Interesting choice, why not simply raise exception?

File: src/module_1/module_1_meteo_api.py [hunk 10]
```diff
@@ -1,7 +1,248 @@
...
+    :return: None
+    """
+
+    # Create a mapping from short variable names to full names
+    short_to_full = {v.split("_")[0]: v for v in VARIABLES}
+
+    # Retrieve the full name of the variable
+    full_variable_name = short_to_full.get(variable, variable)
+
+    plt.figure(figsize=(12, 8))
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: consider exploring `fig, ax = plt.subplots()` since can give mroe control and be more explicit

File: src/module_1/module_1_meteo_api.py [hunk 11]
```diff
@@ -1,7 +1,248 @@
...
+                logging.info(f"Getting data for {model} in {city}")
+                data = get_data_meteo_api(
+                    city, "1950-01-01", "1960-12-31", variable, model
+                )
+                logging.info(data.head())
+                list_data_variables.append(data)
+        combined_df = list_data_variables[0]
+        for i, df in enumerate(list_data_variables[1:], 1):
+            combined_df = pd.merge(
+                combined_df, df, on=["time", "city"], suffixes=("", f"_df{i}")
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py]: This could be done more elegantly. The city is the same for all. You could set time as index for example and use concat. Or concat axis=0 and then pivot.

File: src/module_1/module_1_meteo_api.py [hunk 12]
```diff
@@ -1,7 +1,248 @@
...
+    df["city"] = city
+    df["time"] = pd.to_datetime(df["time"])
+    return df
+
+
+def calculate_mean_std(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Calculates the mean and standard deviation for each prefix in the \
+    DataFrame's columns.
+    :param df: The DataFrame to calculate the mean and standard deviation for.
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:151]: This tells hardly anything. How this dataframe looks like? Which columns do we expect?

File: tests/module_1/test_meteo_api.py
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:1]: Nice! But I would suggest to explore pytest library as well. Personally, I find it easier

File: tests/module_1/test_meteo_api.py [hunk 13]
```diff
@@ -1,7 +1,77 @@
...
+    def test_calculate_mean_std(self):
+        data = {
+            "time": ["2022-01-01", "2022-01-02"],
+            "temperature_2m_AAAAAAAA": [10, 20],
+            "temperature_2m_BBBBBBBB": [15, 25],
+            "city": ["City1", "City1"],
+        }
+        df = pd.DataFrame(data)
+        result = calculate_mean_std(df)
+        self.assertIn("temperature_mean", result.columns)
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:40]: In pandas you have testing methods in pd.testing.assert_[frame|series|...]

File: tests/module_1/test_meteo_api.py [hunk 14]
```diff
@@ -1,7 +1,77 @@
...
+        self.assertIn("temperature_std", result.columns)
+        self.assertEqual(result["temperature_mean"][0], 12.5)
+        self.assertEqual(result["temperature_mean"][1], 22.5)
+
+
+class TestCallAPI(unittest.TestCase):
+    @patch("src.module_1.module_1_meteo_api.requests.get")
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:47]: nice use of patch!

File: tests/module_1/test_meteo_api.py [hunk 15]
```diff
@@ -1,7 +1,77 @@
...
+        # Mock a failed API response
+        mock_response = Mock()
+        mock_get.return_value = mock_response
+        mock_response.status_code = 404
+
+        result = call_api({})
+        self.assertEqual(result.status_code, 404)
 
 
 def test_main():
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py]: I would not test main as unit test.

File: src/module_1/Precipitation_sum.png
Comment (pabaldonedo) [inline src/module_1/Precipitation_sum.png:1]: This plot is hard to follow due to the overlay. It is hard for general population, even more for color blinded (8% of males & 0.5% females). IF you see this you can reconsider other plots, like dashed or pointed lines instead of shadowed areas

**Review by pabaldonedo — APPROVED** [review]
Quite nice work. I really loved the type hinting, the docstrings and style in general. There are some things that could have been done more elegantly (You have teachers solutions to get some inspiration for the future).

The greatest shortcoming is that you were calling the API for each variable and model independently instead in a single call, which leads to inefficient code and too many calls to the API.

I suggest some explorations to increase your toolbox, such as the use of pytest.

Keep the good work! :)

File: src/module_1/annual_data.csv
Comment (Alejandro-Andreo) [reply src/module_1/annual_data.csv:1]: Ah, esto lo hacia básicamente por el rate limit. Como quería testear y cambiar como se veía el plot, guardaba los datos y en vez de volver a hacer api requests, importaba los datos directamente desde local y comentaba todo lo anterior.

File: src/module_1/module_1_meteo_api.py (hunk 5 above)
Comment (Alejandro-Andreo) [reply src/module_1/module_1_meteo_api.py:244]: Sí, en principio, lo hacía todo de una o al menos sin tantas iteraciones. Pero como el rate limit me daba tantos problemas, al final lo he desagregado en muchas calls de poco contenido para ir viendo como lidiar con el limite de la API.

File: src/module_1/module_1_meteo_api.py (hunk 12 above)
Comment (Alejandro-Andreo) [reply src/module_1/module_1_meteo_api.py:151]: Especificar algo como esto?:
So, if the DataFrame has columns 'temperature_2m_mean_df0', 'temperature_2m_mean_df1', 'precipitation_sum_df0', 'precipitation_sum_df1', 'soil_moisture_0_to_10cm_mean_df0', and 'soil_moisture_0_to_10cm_mean_df1', the function will calculate the mean and standard deviation for the 'temperature_2m_mean', 'precipitation_sum', and 'soil_moisture_0_to_10cm_mean' prefixes. Once the mean and standard deviation are calculated, the original columns are dropped and the new columns are added to the DataFrame. Other columns without underscores are left untouched(i.e. 'city' and 'time')

File: src/module_1/module_1_meteo_api.py (hunk 9 above)
Comment (Alejandro-Andreo) [reply src/module_1/module_1_meteo_api.py:66]: La verdad que no sabía muy bien que hacer cuando había un error al llamar a la API, si hacer un raise systemExit() o retornar un None.

File: src/module_1/module_1_meteo_api.py (hunk 8 above)
Comment (Alejandro-Andreo) [reply src/module_1/module_1_meteo_api.py]: Ah, gracias por darte cuenta, es una errata. Debería retornar None, pues solo retorna la response en caso de que pase la función de validación.

File: src/module_1/Precipitation_sum.png
Comment (Alejandro-Andreo) [reply src/module_1/Precipitation_sum.png:1]: He intentado cambiar como se ven los gráficos. Ahora, a parte del color para distinguir las distintas ciudades, sus medias se identifican por marcadores: Circulo, cuadrado y diamante. He quitado el _fill-in-between_ de los limites superiores e inferiores y lo he sustituido por _dashed_ y _dashdot lines_. También los he puesto en la leyenda. 
soil_moisture

**Comment by Alejandro-Andreo** [comment]
He hecho unos cuentos cambios en base a lo que me has dicho. Dime que opinas y si ya está todo correcto, cierro y hago merge a la pull request.

File: src/module_1/module_1_meteo_api.py (hunk 11 above)
Comment (Alejandro-Andreo) [reply src/module_1/module_1_meteo_api.py]: Lo he cambiado a esto:
```python
combined_df = pd.concat([df.set_index(['time', 'city']) for df in list_data_variables], axis=1)
```
Como hago las API calls de en una en una, todas las columnas con las misma variables se llaman igual así que tengo que renombrar a las columnas y luego ya resetear el índice.
```python
combined_df.columns = [f'{col}_{i}' if i > 0 else col for i, col in enumerate(combined_df.columns)]
combined_df = combined_df.reset_index()
```

File: src/module_1/module_1_meteo_api.py (hunk 12 above)
Comment (pabaldonedo) [reply src/module_1/module_1_meteo_api.py:151]: I normally do the following:

    `df: a nxm DataFrame containing data about X with columns [col1, col2, col3]  and whatever as indices`

Just by reading that I can now what to expect in terms of shape, content and naming

File: src/module_1/Precipitation_sum.png
Comment (pabaldonedo) [reply src/module_1/Precipitation_sum.png:1]: ¡Creo que es mucho más legible! Un consejo adicional (igual lo has seguido) se puede elegir paletas de color que sean más robustas contra daltónicos para hacer el gráfico más útil para más gente. Un ejemplo de webs que te ayudan con elegir la paleta es [esta](https://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3)

File: src/module_1/module_1_meteo_api.py [hunk 16]
```diff
@@ -166,43 +181,62 @@ def plot_data(df: pd.DataFrame, variable: str) -> None:
...
-            marker="o",
+            marker=markers[idx % len(markers)],
             linestyle="-",
             linewidth=2,
+            markersize=8,
         )
-        plt.fill_between(
-            city_data["year"], mean - std, mean + std, color=colors[idx], alpha=0.3
+        # Plot standard deviation with dashed lines
+        plt.plot(
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:205]: the point of fig, ax = plt.subplots() is to use "ax.plot()" here to gain control of whre to plot things in matplotlib. Plt will take the "current axes" but having the axes object ax we explicitly tells matplotlib where to plot.

File: src/module_1/module_1_meteo_api.py [hunk 17]
```diff
@@ -166,43 +181,62 @@ def plot_data(df: pd.DataFrame, variable: str) -> None:
...
+            mean - std,
+            label=f"{city} Lower bound",
+            color=colors[idx],
+            linestyle="dashdot",
+            linewidth=2,
+            alpha=0.7,
         )
 
-    plt.title(
+    ax.set_title(
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:224]: Here you go with the ax!

**Review by pabaldonedo — APPROVED** [review]
Nice PR! You are more than welcome to merge it!
